### PR TITLE
Add TokenClassification for Mistral, Mixtral and Qwen2

### PR DESCRIPTION
**Pull request by josephenguehard**
# What does this PR do?

This PR adds token classification for Llama and its extensions, where code from Llama was copied : Gemma, Mistral, Mixtral, Persimmon, Qwen2, Qwen2MoE, StableLm and StarCoder2:

- `GemmaForTokenClassification`
- `LlamaForTokenClassification`
- `MistralForTokenClassification`
- `MixtralForTokenClassification`
- `PersimmonForTokenClassification`
- `Qwen2ForTokenClassification`
- `Qwen2MoeForTokenClassification`
- `StableLmForTokenClassification`
- `StarCoder2ForTokenClassification`

Note: The original PR was intended for Mistral only. Llama has been added as a request from @KoichiYasuoka, and the others models have been added as copying most of Llama setting.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

- text models: @ArthurZucker and @younesbelkada

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks for the PR, could you explain the motivation behind this? No checkpoints are publicly available, and we usually first oopen a feature request and then integrate it if the community shows interest if there are no official checkpoints!

**Comment by KoichiYasuoka** [comment]
Thank you @josephenguehard for 

* `MistralForTokenClassification`

since I've already released https://huggingface.co/KoichiYasuoka/Swallow-MS-7b-upos (using `trust_remote_code=True`). And would you kindly include

* `LlamaForTokenClassification`

from #26521 for my https://huggingface.co/KoichiYasuoka/Swallow-7b-plus-upos ?

**Comment by josephenguehard** [comment]
Hi Both! 

Thank you for your replies! I will open a separate issue to discuss adding these classes.

@KoichiYasuoka I'd be happy to add `LlamaForTokenClassification`! I'll try to have this done by the end of the week.

**Comment by josephenguehard** [comment]
The feature request issue linked to this PR is #29940

**Comment by josephenguehard** [comment]
@KoichiYasuoka I've added Llama support as well as its extensions (models where code from Llama was copied), and I think the PR is ready for review.

One point where I'd like to have a discussion is on the Dropout. I see in [your implementation](https://github.com/huggingface/transformers/issues/26521#issuecomment-1868284434) that you include a Dropout layer, which follows the [BertForTokenClassification](https://github.com/huggingface/transformers/blob/17b06e2c6650de162e7954babf6224c1975c2852/src/transformers/models/bert/modeling_bert.py#L1691).
However, I argue that this implementation actually follows [BertForSequenceClassification](https://github.com/huggingface/transformers/blob/17b06e2c6650de162e7954babf6224c1975c2852/src/transformers/models/bert/modeling_bert.py#L1494), which also has a Dropout layer. Moreover, [LlamaForSequenceClassification](https://github.com/huggingface/transformers/blob/17b06e2c6650de162e7954babf6224c1975c2852/src/transformers/models/llama/modeling_llama.py#L1343) does not have such Dropout layer, so I think we should not include it for `LlamaForTokenClassification` either. But happy to get other's opinions!

**Comment by KoichiYasuoka** [comment]
Well, Dropout layer... My Implementation follows [GPT2ForTokenClassification](https://github.com/huggingface/transformers/blob/v4.10.0/src/transformers/models/gpt2/modeling_gpt2.py#L1344) which was released with #13290 at Transformers [v4.10.0](https://github.com/huggingface/transformers/releases/tag/v4.10.0). But I'm vague how GPT2 and LLaMA differ at that point...

**Comment by josephenguehard** [comment]
It's very easy for me to add Dropout if there is a consensus on it. 
FYI I'm away and won't have access to my laptop until April 15th, will pick this up then.

**Comment by KoichiYasuoka** [comment]
Yesterday @SeanLee97 of @WhereIsAI released two models:

* https://huggingface.co/WhereIsAI/billm-llama-7b-conll03-ner
* https://huggingface.co/WhereIsAI/billm-mistral-7b-conll03-ner

They use their own package [BiLLM](https://pypi.org/project/BiLLM/) with [LlamaForTokenClassification](https://github.com/WhereIsAI/BiLLM/blob/main/src/billm/modeling_llama.py#L230) and [MistralForTokenClassification](https://github.com/WhereIsAI/BiLLM/blob/main/src/billm/modeling_mistral.py#L204).

**Comment by SeanLee97** [comment]
> Yesterday @SeanLee97 of @WhereIsAI released two models:
> 
> * https://huggingface.co/WhereIsAI/billm-llama-7b-conll03-ner
> * https://huggingface.co/WhereIsAI/billm-mistral-7b-conll03-ner
> 
> They use their own package [BiLLM](https://pypi.org/project/BiLLM/) with [LlamaForTokenClassification](https://github.com/WhereIsAI/BiLLM/blob/main/src/billm/modeling_llama.py#L230) and [MistralForTokenClassification](https://github.com/WhereIsAI/BiLLM/blob/main/src/billm/modeling_mistral.py#L204).

Thank you for mentioning our released models. Happy to see that official 🤗 transformers supports token classification. 
However, based on our experiments in the paper (https://arxiv.org/abs/2310.01208), LLMs with the causal mask cannot perform well on token classification tasks. Thus, we convert the causal attention mask from uni- to bi-directional. This change can improve the token classification performance significantly. Hopefully this feature can be added to the official transformers in the future.

**Comment by KoichiYasuoka** [comment]
I'm waiting for your review @ArthurZucker

File: tests/models/gemma/test_modeling_gemma.py [hunk 1]
```diff
@@ -367,6 +372,21 @@ def test_Gemma_sequence_classification_model_for_multi_label(self):
         result = model(input_ids, attention_mask=attention_mask, labels=sequence_labels)
         self.assertEqual(result.logits.shape, (self.model_tester.batch_size, self.model_tester.num_labels))
 
+    def test_Gemma_token_classification_model(self):
```
Comment (ArthurZucker) [inline tests/models/gemma/test_modeling_gemma.py:375]: tests can also be copied from!

**Review by ArthurZucker — APPROVED** [review]
Thanks! Let's just add copied from for the test and good to go. 
Seems to have decent traction, which is what I was waiting for!

**Comment by ArthurZucker** [comment]
alsor make sure to resolve the conflics

**Comment by josephenguehard** [comment]
@ArthurZucker I've added `Copied from` for the tests and resolved the conflicts.

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -1540,3 +1541,87 @@ def forward(
...
+)
+class LlamaForTokenClassification(LlamaPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+        self.num_labels = config.num_labels
+        self.model = LlamaModel(config)
+        if hasattr(config, "classifier_dropout") and config.classifier_dropout is not None:
+            classifier_dropout = config.classifier_dropout
+        elif hasattr(config, "hidden_dropout") and config.hidden_dropout is not None:
+            classifier_dropout = config.hidden_dropout
```
Comment (younesbelkada) [inline src/transformers/models/llama/modeling_llama.py:1561]:
```suggestion
        if getattr(config, "classifier_dropout", None) is not None:
            classifier_dropout = config.classifier_dropout
        elif getattr(config, "hidden_dropout", None) is not None:
            classifier_dropout = config.hidden_dropout
```

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot ! LGTM with only one nit - can you propagate these changes to all models with `make fix-copies` (once the suggestion accepted)

**Comment by josephenguehard** [comment]
Done @younesbelkada 👍

**Review by younesbelkada — APPROVED** [review]
Clean, thanks !
